src/mlpack/core/tree/cosine_tree/cosine_tree_impl.hpp

```diff
@@ -0,0 +1,92 @@
+/**
+ * @file cosine_tree_impl.hpp
+ * @author Sriram S K
+ *
+ * Implementation of cosine tree.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_TREE_COSINE_TREE_COSINE_TREE_IMPL_HPP
+#define MLPACK_CORE_TREE_COSINE_TREE_COSINE_TREE_IMPL_HPP
+
+#include "cosine_tree.hpp"
+#include <mlpack/core/util/log.hpp>
+
+namespace mlpack {
+namespace tree {
+
+/**
+ * Serialize the tree.
+ */
+template<typename Archive>
+void CosineTree::serialize(Archive& ar, const unsigned int /* version */)
+{
+  // If we're loading, and we have children, they need to be deleted.
+  if (Archive::is_loading::value)
+  {
+    if (left)
+      delete left;
+    if (right)
+      delete right;
+    if (!parent)
+      delete dataset;
```

Comment:
Does this deletion need to happen only if `localDataset` is true?

Reply:
Should be `!parent && localDataset` correct?